nvhpc building block with C++20 library support

According to the [nvhpc docs](https://docs.nvidia.com/hpc-sdk/hpc-sdk-release-notes/index.html#platform-requirements), one needs gcc/glibc>=10.1 for C++20 library support in the new nvhpc 22.3. 

Neither
```
Stage0 += baseimage(image='nvcr.io/nvidia/nvhpc:22.3-devel-cuda11.6-ubuntu20.04')
Stage0 += gnu(version='10')
```
nor 
```
Stage0 += baseimage(image='ubuntu:20.04')
Stage0 += gnu(version='10')
Stage0 += nvhpc(eula=True, cuda_multi=False, version='22.3')
```
seem to do the trick. The second one even actively installs gcc 9 again instead of using the provided gcc 10 toolchain. 
I tested the support by trying to compile a source file with `#include <ranges>` via `nvc++ -std=c++20` inside the container (which works when using the installed `g++` 10.3.0). 

Maybe the nvhpc building block needs a `toolchain` argument?

I have now found out that gcc 10 is not enough for full C++20 support, as `__cplusplus` [is too low](https://stackoverflow.com/questions/71036685/using-c20-with-gcc-10-is-limited-to-c17). `nvc++` uses this macro to determine how it evaluates iterators to be random access iterators. With the C++17 rules that it will follow wih gcc 10, a transform view does not provide random access iterators, but with C++20 rules it does (new concept). So to use a transform view (and probably others as well) in a parallel algorithm, one needs gcc 11. I tried installing it from source using the building block
```python
Stage0 += gnu(version='11.2.0', source=True)
```
but got 
```shell
checking LIBRARY_PATH variable... contains current directory
configure: error: 
*** LIBRARY_PATH shouldn't contain the current directory when
*** building gcc. Please change the environment variable
*** and run configure again.
```
Should I file a new issue? Or is this just an oversight on my side (e.g. singularity build flags)?

--- a/docs/building_blocks.md
+++ b/docs/building_blocks.md
@@ -3163,6 +3163,10 @@ wildcards are supported.  The default is an empty list.
 be installed by downloading the tar package file.  If False,
 install from the package repository.  The default is False.
 
+- __toolchain__: The toolchain object to be used to configure the HPC
+SDK with a specific GNU toolchain.  The default is empty, i.e., use
+the default GNU toolchain.
+
 - __url__: The location of the package that should be installed.  The default value is `https://developer.download.nvidia.com/hpc-sdk/nvhpc_X_Y_Z_cuda_multi.tar.gz`, where `X, `Y`, and `Z` are the year, version, and architecture whose values are automatically determined.
 
 - __version__: The version of the HPC SDK to use.  Note when `package`

--- a/hpccm/building_blocks/nvhpc.py
+++ b/hpccm/building_blocks/nvhpc.py
@@ -107,6 +107,10 @@ class nvhpc(bb_base, hpccm.templates.downloader, hpccm.templates.envvars,
     be installed by downloading the tar package file.  If False,
     install from the package repository.  The default is False.
 
+    toolchain: The toolchain object to be used to configure the HPC
+    SDK with a specific GNU toolchain.  The default is empty, i.e., use
+    the default GNU toolchain.
+
     url: The location of the package that should be installed.  The default value is `https://developer.download.nvidia.com/hpc-sdk/nvhpc_X_Y_Z_cuda_multi.tar.gz`, where `X, `Y`, and `Z` are the year, version, and architecture whose values are automatically determined.
 
     version: The version of the HPC SDK to use.  Note when `package`
@@ -167,6 +171,7 @@ class nvhpc(bb_base, hpccm.templates.downloader, hpccm.templates.envvars,
         self.__redist = kwargs.get('redist', [])
         self.__stdpar_cudacc = kwargs.get('stdpar_cudacc', None)
         self.__tarball = kwargs.get('tarball', False)
+        self.__toolchain = kwargs.get('toolchain', None)
         self.__url = kwargs.get('url', None)
         self.__version = kwargs.get('version', '22.3')
         self.__wd = kwargs.get('wd', hpccm.config.g_wd) # working directory
@@ -236,6 +241,23 @@ class nvhpc(bb_base, hpccm.templates.downloader, hpccm.templates.envvars,
                 apt_repositories=['deb [trusted=yes] https://developer.download.nvidia.com/hpc-sdk/ubuntu/{} /'.format(self.__arch_label)],
                 ospackages=[self.__nvhpc_package],
                 yum_repositories=['https://developer.download.nvidia.com/hpc-sdk/rhel/nvhpc.repo'])
+
+        if self.__toolchain:
+            # Regenerate the localrc using the compilers from the specified
+            # toolchain
+            compiler_bin = posixpath.join(self.__basepath, 'compilers', 'bin')
+
+            args = ['-x']
+            if self.__toolchain.CC:
+                args.append('-gcc {}'.format(self.__toolchain.CC))
+            if self.__toolchain.CXX:
+                args.append('-gpp {}'.format(self.__toolchain.CXX))
+            if self.__toolchain.F77:
+                args.append('-g77 {}'.format(self.__toolchain.F77))
+
+            self += shell(commands=['{0} {1} {2}'.format(
+                posixpath.join(compiler_bin, 'makelocalrc'), compiler_bin,
+                ' '.join(args))])
 
         self += environment(variables=self.environment_step())
 

--- a/test/test_nvhpc.py
+++ b/test/test_nvhpc.py
@@ -25,6 +25,7 @@ import unittest
 from helpers import aarch64, centos, docker, ppc64le, ubuntu, x86_64
 
 from hpccm.building_blocks.nvhpc import nvhpc
+from hpccm.toolchain import toolchain
 
 class Test_nvhpc(unittest.TestCase):
     def setUp(self):
@@ -240,6 +241,30 @@ COPY --from=0 /opt/nvidia/hpc_sdk/Linux_x86_64/22.3/REDIST/compilers/lib/* /opt/
 COPY --from=0 /opt/nvidia/hpc_sdk/Linux_x86_64/22.3/REDIST/math_libs/11.0/lib64/libcufft.so.10 /opt/nvidia/hpc_sdk/Linux_x86_64/22.3/math_libs/11.0/lib64/libcufft.so.10
 COPY --from=0 /opt/nvidia/hpc_sdk/Linux_x86_64/22.3/REDIST/math_libs/11.0/lib64/libcublas.so.11 /opt/nvidia/hpc_sdk/Linux_x86_64/22.3/math_libs/11.0/lib64/libcublas.so.11
 ENV LD_LIBRARY_PATH=/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/comm_libs/11.0/nccl/lib:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/compilers/lib:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/math_libs/11.0/lib64:$LD_LIBRARY_PATH''')
+
+    @x86_64
+    @ubuntu
+    @docker
+    def test_makelocalrc(self):
+        """makelocalrc"""
+        tc = toolchain(CC='gcc-10', CXX='g++-10', F77='gfortran-10')
+        n = nvhpc(eula=True, mpi=False, toolchain=tc, version='22.3')
+        self.assertMultiLineEqual(str(n),
+r'''# NVIDIA HPC SDK version 22.3
+RUN apt-get update -y && \
+    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+        ca-certificates && \
+    rm -rf /var/lib/apt/lists/*
+RUN echo "deb [trusted=yes] https://developer.download.nvidia.com/hpc-sdk/ubuntu/amd64 /" >> /etc/apt/sources.list.d/hpccm.list && \
+    apt-get update -y && \
+    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+        nvhpc-22-3-cuda-multi && \
+    rm -rf /var/lib/apt/lists/*
+RUN /opt/nvidia/hpc_sdk/Linux_x86_64/22.3/compilers/bin/makelocalrc /opt/nvidia/hpc_sdk/Linux_x86_64/22.3/compilers/bin -x -gcc gcc-10 -gpp g++-10 -g77 gfortran-10
+ENV CPATH=/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/comm_libs/nvshmem/include:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/comm_libs/nccl/include:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/compilers/extras/qd/include/qd:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/math_libs/include:$CPATH \
+    LD_LIBRARY_PATH=/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/comm_libs/nvshmem/lib:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/comm_libs/nccl/lib:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/math_libs/lib64:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/compilers/lib:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/cuda/lib64:$LD_LIBRARY_PATH \
+    MANPATH=/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/compilers/man:$MANPATH \
+    PATH=/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/comm_libs/nvshmem/bin:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/comm_libs/nccl/bin:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/profilers/bin:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/compilers/bin:/opt/nvidia/hpc_sdk/Linux_x86_64/22.3/cuda/bin:$PATH''')
 
     def test_toolchain(self):
         """Toolchain"""

